libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,693 @@
+
+    memset(_id_or_mac,0,ODID_ID_SIZE);
+    memset(_uas_id,0,ODID_ID_SIZE + 1);
+    memset(_operator_id,0,ODID_ID_SIZE + 1);
+    memset(_description,0,ODID_STR_SIZE + 1);
+
+    transmit_basic_ID_message = false; //default we don't send any packets
+    transmit_location_message = false;
+    transmit_self_ID_message = false;
+    transmit_system_message = false;
+    transmit_operator_id_message = false;
+
+    // UA_Type Parameter
+    // If it has been set then use it else pre-arm error will be set by the arming checks
+    _ua_type = MAV_ODID_UA_TYPE_NONE;
+    if (_ua_type_parm.configured()) {
+        _ua_type = MAV_ODID_UA_TYPE(_ua_type_parm.get());
+    }
+
+//code to be replaced: set dummy values in open drone ID messages
+    set_mavlink_channel(MAVLINK_COMM_1); //set to MAVLink channel 1
+    set_mavlink_dynamic_messages_period_ms(1000); //send dynamic messages at 1 Hz frequency i.e. every 1 second
+    set_mavlink_static_messages_period_ms(10000); //send dynamic messages at 0.1 Hz frequency i.e. every 10 seconds
+    set_id_type(MAV_ODID_ID_TYPE_SERIAL_NUMBER);
+
+    set_operator_location_type(MAV_ODID_OPERATOR_LOCATION_TYPE_LIVE_GNSS);
+    set_classification_type(MAV_ODID_CLASSIFICATION_TYPE_UNDECLARED);
+    set_category_eu(MAV_ODID_CATEGORY_EU_UNDECLARED);
+    set_class_eu(MAV_ODID_CLASS_EU_UNDECLARED);
+    set_operator_id_type(MAV_ODID_OPERATOR_ID_TYPE_CAA);
+    set_description_type(MAV_ODID_DESC_TYPE_TEXT);
+
+    set_operator_latitude(10);
+    set_operator_longitude(30);
+    set_operator_timestamp((time_t)(AP_HAL::micros())/1000); //set to current time
+    set_area_count(1); //default 1
+    set_area_radius(0); //default 0
+    set_area_ceiling(-1000); //default -1000
+    set_area_floor(-1000); //default -1000
+    set_operator_altitude_geo(-1000); //default -1000
+
+    uint8_t id_or_mac[ODID_ID_SIZE] = "000000000000";
+    set_id_or_mac(id_or_mac);
+
+    char uas_id[ODID_ID_SIZE + 1] = "SN000000123456789";
+    set_uas_id(uas_id);
+
+    char operator_id[ODID_ID_SIZE + 1] = "NLD000000123456789";
+    set_operator_id(operator_id);
+
+    char description[ODID_STR_SIZE + 1] = "MAVLink test flight";
+    set_description(description);
+
+    transmit_basic_ID_message = true;  //broadcast all message types
+    transmit_location_message = true;
+    transmit_self_ID_message = true;
+    transmit_system_message = true;
+    transmit_operator_id_message = true;
+//end dummy values
+
+}
+
+// Perform the pre-arm checks and prevent arming if they are not satisifed
+// Except in the case of an in-flight reboot
+bool AP_OpenDroneID::pre_arm_check(char* failmsg, uint8_t failmsg_len) const
+{
+    // UA_Type Parameter
+    // If it has been set then use it else pre-arm error will be set by the arming checks
+    if (_ua_type_parm.get() == MAV_ODID_UA_TYPE_NONE) {
+        // Set Pre-arm error
+        strncpy(failmsg, "UA_TYPE Must Not Be NONE", failmsg_len);
+        return false;
+    }
+
```

Comment:
I think it would make sense to also sanity check here that the serial number of the UA is non-zero.

Reply:
sanity checks like that should be in the transmitter module, which gives go/no-go via ARM_STATUS message

Reply:
we check for gps fix before using the gps data